### PR TITLE
Whitelist, stickybans and db and admin stuff

### PR DESCRIPTION
**Pull request by Rohesie**
Also moved several procs towards locations to match the /tg/ ones.
Will need a review from @psykzz or someone who knows better about the database stuff.

### CODE REVIEW

File: code/modules/admin/IsBanned.dm [hunk 1]
```diff
@@ -0,0 +1,233 @@
+//Blocks an attempt to connect before even creating our client datum thing.
+
+//How many new ckey matches before we revert the stickyban to it's roundstart state
+//These are exclusive, so once it goes over one of these numbers, it reverts the ban
+#define STICKYBAN_MAX_MATCHES 15
```
Comment (spookydonut) [inline code/modules/admin/IsBanned.dm]: macros defined twice

**Comment by spookydonut** [comment]
```
code/modules/admin/IsBanned.dm, line 5, column 9:
warning: macro redefined: STICKYBAN_MAX_MATCHES
- code/__DEFINES/admin.dm:104:9: previous definition of STICKYBAN_MAX_MATCHES

code/modules/admin/IsBanned.dm, line 6, column 9:
warning: macro redefined: STICKYBAN_MAX_EXISTING_USER_MATCHES
- code/__DEFINES/admin.dm:105:9: previous definition of STICKYBAN_MAX_EXISTING_USER_MATCHES

code/modules/admin/IsBanned.dm, line 7, column 9:
warning: macro redefined: STICKYBAN_MAX_ADMIN_MATCHES
- code/__DEFINES/admin.dm:106:9: previous definition of STICKYBAN_MAX_ADMIN_MATCHES
```

File: code/controllers/subsystem/dbcore.dm [hunk 2]
```diff
@@ -63,7 +65,11 @@ SUBSYSTEM_DEF(dbcore)
 	if(IsConnected())
 		return TRUE
 
-	if(failed_connections > FAILED_DB_CONNECTION_CUTOFF)	//If it failed to establish a connection more than 5 times in a row, don't bother attempting to connect anymore.
+	if(failed_connection_timeout <= world.time) //it's been more than 5 seconds since we failed to connect, reset the counter
+		failed_connections = 0
+
+	if(failed_connections > FAILED_DB_CONNECTION_CUTOFF)	//If it failed to establish a connection more than 5 times in a row, don't bother attempting to connect for 5 seconds.
+		failed_connection_timeout = world.time + 50
```
Comment (psykzz) [inline code/controllers/subsystem/dbcore.dm]:
```suggestion
		failed_connection_timeout = world.time + 5 SECONDS
```

File: code/modules/admin/IsBanned.dm [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+			else
+				log_access("Failed Login: [key] - Not on whitelist")
+				return list("reason"="whitelist", "desc" = "\nReason: You are not on the white list for this server")
+
+	//Guest Checking
+	if(!real_bans_only && !C && IsGuestKey(key))
+		if (CONFIG_GET(flag/guest_ban))
+			log_access("Failed Login: [key] - Guests not allowed")
+			return list("reason"="guest", "desc"="\nReason: Guests not allowed. Please sign in with a byond account.")
+		if (CONFIG_GET(flag/panic_bunker) && SSdbcore.Connect())
```
Comment (psykzz) [inline code/modules/admin/IsBanned.dm]: Your connecting to the database here before doing sql_enabled check.
Is that okay?

Comment (Rohesie) [reply code/modules/admin/IsBanned.dm]: It's not ideal, there's quite a bit I'd clean up on this proc, but this is /tg/ code and I prefer compatibility over clarity in this case. So when they improve stuff we'll have less work porting. This is not code we'll be accessing often anwyay.
I guess it means that the `flag/panic_bunker` assumes `flag/sql_enabled` to be true.

**Comment by Rohesie** [comment]
This one should be ready for merge.

**Pull request merged**